### PR TITLE
Fix:k8s liveness probe fails during pod termination

### LINKED ISSUE #107473: k8s liveness probe fails during pod termination
### What happened?

I've tested with k8s v1.21 on Amazon EKS and k8s v1.22, v1.21, and v1.20 in Minikube

I installed a pod containing a simple http server having some `/alive` endpoint.  In the pod spec, specify a liveness probe that uses this endpoint.
No liveness probe failure while pod is running, as expected.
But after executing a `kubectl delete po` command, will see k8s event that liveness probe failed.

I can reliably reproduce this.  tried with servers written in spring boot, golang, and python/gunicorn

Then I modified the server, added a second endpoint `/slow` that sleeps for 30 seconds
I call this endpoint, then immediately execute `kubectl delete po` then `kubectl get po -o yaml`
Can see that deletionTimestamp is now non-null, but liveness probe continually fails while app spends 30 seconds trying to shut-down

Starting minikube command `minikube start --extra-config=kubelet.v=4`, I never see log `Pod deletion requested, setting probe result to success` from https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/prober/worker.go#L234

### What did you expect to happen?

Based on [this](https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/kubelet/prober/worker.go#L233) logic, I would expect liveness probe not to fail during shutdown

### How can we reproduce it (as minimally and precisely as possible)?

Install a pod containing an http server.  Server should expose an endpoint that's used by livenessProbe in the Pod spec
Delete the pod.
Now, execute `kubectl get events`, see the liveness probe failure event(s)

for example,

```
apiVersion: v1
kind: Pod
metadata:
  name: test-server
spec:
  containers:
    - name: test-server
      image: waltersmatthew/goserver:latest
      ports:
        - name: http
          containerPort: 8080
          protocol: TCP
      livenessProbe:
        httpGet:
          path: /alive
          port: http
        periodSeconds: 1
```

### Anything else we need to know?

_No response_

### Kubernetes version

1.21.0, 1.21.2 among others

### Cloud provider

AWS EKS, minikube

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by 249043822**
#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Fixes #107473

#### Special notes for your reviewer:
/cc @matthyx @yangjunmyfm192085 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/prober/common_test.go
Comment (yangjunmyfm192085) [inline pkg/kubelet/prober/common_test.go:111]: Yeah, the configsource always sends a copy of the pod to config source channel,so we added podManager here.

**Review by yangjunmyfm192085 — COMMENTED** [review]
/lgtm
/cc @mrunalp @ehashman @matthyx

**Comment by matthyx** [comment]
@249043822 could you add some test coverage of the bug to prevent regression?

**Comment by matthyx** [comment]
> @249043822 could you add some test coverage of the bug to prevent regression?

I think e2e test is needed here, because the bug I introduced in https://github.com/kubernetes/kubernetes/pull/98571 slipped through because I was using a mock `probeManager`.
Good job here!
/lgtm
/hold
(holding for the missing test coverage of that specific bug)

File: test/e2e/common/node/container_probe.go
Comment (matthyx) [inline test/e2e/common/node/container_probe.go:568]: The description doesn't quite match the actual test... what you are testing is that the container doesn't become unhealthy (indicating a livenessProbe failure) once the pod deletion has been requested (graceful shutdown indeed).

**Review by matthyx — CHANGES_REQUESTED** [review]
Also a little trick when you're doing TDD, you can split your changes into 2 commits: first the e2e test (to prove it fails with existing code, illustrating the bug) and second your fix (to prove it's fixed).

**Comment by 249043822** [comment]
> 



> Also a little trick when you're doing TDD, you can split your changes into 2 commits: first the e2e test (to prove it fails with existing code, illustrating the bug) and second your fix (to prove it's fixed).

@matthyx thanks for your advice, this bug only has a little certain probability, because the housekeeping may clean the probe before it has chance to execute again, https://github.com/kubernetes/kubernetes/blob/7dd03c8326858ceb232d38476db3240a15e8494f/pkg/kubelet/kubelet_pods.go#L1106 it's hard to write a e2e test to prove the fail

**Review by yangjunmyfm192085 — COMMENTED** [review]
/lgtm

**Comment by matthyx** [comment]
> it's hard to write a e2e test to prove the fail

Indeed, I see the issue...
Ok so after some local running, I can confirm the e2e test nails the bug, and the code change fixes it.
Well done!
/lgtm

**Review by yangjunmyfm192085 — COMMENTED** [review]
also need
/assign @ehashman @SergeyKanzhelev

File: pkg/kubelet/prober/worker.go
Comment (smarterclayton) [inline pkg/kubelet/prober/worker.go:233]: There's an overlapping change that needs to be made before this will be totally correct - you can't rely on the pod manager to correctly indicate whether a pod "has terminated", because the pod manager just tracks desired pods, not "pods the kubelet is running".

There is a separate change https://github.com/kubernetes/kubernetes/issues/105014 that probably needs to be done before this merges to more correctly take into account how pod lifecycle must handle probes.

However, even after that change this may not be the correct behavior, because a pod that has defined a liveness probe AND expects a long grace period may specify a shorter termination period for liveness failures (i.e. a container that may take 10 minutes to clean up is allowed to define a liveness probe that if fails, shuts the pod down in 10 seconds).  For more detail on that, see 

    kubectl explain deployment.spec.template.spec.containers.livenessProbe.terminationGracePeriodSeconds

**Comment by smarterclayton** [comment]
I want to address this after #105014 lands

File: pkg/kubelet/prober/worker.go
Comment (249043822) [reply pkg/kubelet/prober/worker.go:233]: This change that relies on pod manager is only to indicate whether a pod "has deleted",  not "has terminated".
As the housekeeping still would shutdown probes for undesired pods, https://github.com/kubernetes/kubernetes/blob/40c2d049465417f510e4182b05953a49fc5693d4/pkg/kubelet/kubelet_pods.go#L1106

Actually if we consider the feature ProbeTerminationGracePeriod, this logic above for probe is not correct, I just want to fix that bug https://github.com/kubernetes/kubernetes/issues/107473 at that time. Now I think these codes may be not useful, actually from our tests, the probes for terminating(deteted) pod are basically always shutted down by housekeeping, the codes above have no chance to run.

**Comment by endocrimes** [comment]
https://github.com/kubernetes/kubernetes/issues/105014 has landed.

Sorry for the ping with new job stuff going on, but @smarterclayton plz could you take a look?

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by rphillips** [comment]
I do not believe we need this anymore. Liveness probes are stopped during pod termination via

https://github.com/kubernetes/kubernetes/pull/110191

/close

**Pull request closed without merging**
